contrib/mesos/pkg/queue/delay/indexed_delay_queue.go

```diff
@@ -0,0 +1,248 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package delay
+
+import (
+	"container/heap"
+	"fmt"
+	"reflect"
+	"time"
+
+	"github.com/pivotal-golang/clock"
+
+	"k8s.io/kubernetes/contrib/mesos/pkg/queue"
+	"k8s.io/kubernetes/contrib/mesos/pkg/queue/priority"
+	"k8s.io/kubernetes/pkg/util/sets"
+)
+
+// IndexedDelayQueue is a thread-safe, time-based queue with uniquely identified items.
+//
+// Adding or offering an item with an ID matching an item still in the queue will
+// invoke the configured SchedulingPolicy and supplied ValueReplacementPolicy to
+// determine the desired behavior.
+//
+// Items can only be popped after their event time has been reached.
+// Use Add to specify a delay duration and have the event time calculated.
+// Use Offer to specify a specific event time.
+type IndexedDelayQueue struct {
+	*Queue
```

Comment:
I like that even less than having a named field. However, we have a bigger problem I only just realized. The priority queue requires container/heap to be used correctly, otherwise it's LIFO. The delay queues tho are directly interfaced... So DelayQueue can't "extend" PriorityQueue or it violates liskov substitution. 

So I'll have to use a named field. 

That said, I think IndexedDelayQueue is too tightly coupled with DelayQueue. I might move it to an "indexed" package just to make sure DelayQueue is properly extensible. I can already tell that's gonna hurt, because they're sharing locks... But if I can get it right it should cut down on the code duplication.